File: queries/h19.R
tpc_h19 <- function(input_func, collect_func = dplyr::collect) {
  joined <- input_func("lineitem") %>%
    inner_join(input_func("part"), by = c("l_partkey" = "p_partkey"))

  result <- joined %>%
    filter(
      (
        p_brand == "Brand#12" &
          p_container %in% c('SM CASE', 'SM BOX', 'SM PACK', 'SM PKG') &
          l_quantity >= 1 &
          l_quantity <= (1 + 10) &
          p_size >= 1 &
          p_size <= 5 &
          l_shipmode %in% c("AIR", "AIR REG") &
          l_shipinstruct == "DELIVER IN PERSON"
      ) |
        (
          p_brand == "Brand#23" &
            p_container %in% c('MED BAG', 'MED BOX', 'MED PKG', 'MED PACK') &
            l_quantity >= 10 &
            l_quantity <= (10 + 10) &
            p_size >= 1 &
            p_size <= 10 &
            l_shipmode %in% c("AIR", "AIR REG") &
            l_shipinstruct == "DELIVER IN PERSON"
        ) |
        (
          p_brand == "Brand#34" &
            p_container %in% c('LG CASE', 'LG BOX', 'LG PACK', 'LG PKG') &
            l_quantity >= 20 &
            l_quantity <= (20 + 10) &
            p_size >= 1 &
            p_size <= 15 &
            l_shipmode %in% c("AIR", "AIR REG") &
            l_shipinstruct == "DELIVER IN PERSON"
        )
    )

  result %>%
    summarise(
      revenue = sum(l_extendedprice * (1 - l_discount))
    ) %>%
    collect_func()
}

File: ibis_tpc/h19.py

def tpc_h19(con, QUANTITY1=1, QUANTITY2=10, QUANTITY3=20,
            BRAND1='Brand#12', BRAND2='Brand#23', BRAND3='Brand#34'):
    '''Discounted Revenue Query (Q19)

    The Discounted Revenue Query reports the gross discounted revenue
    attributed to the sale of selected parts handled in a particular manner.
    This query is an example of code such as might be produced programmatically
    by a data mining tool.'''

    lineitem = con.table('lineitem')
    part = con.table('part')
    q = lineitem.join(part, part.p_partkey == lineitem.l_partkey)
    q = q.materialize()

    q1 = ((q.p_brand == BRAND1) &
          (q.p_container.isin(('sm case', 'sm box', 'sm pack', 'sm pkg'))) &
          (q.l_quantity >= QUANTITY1) &
          (q.l_quantity <= QUANTITY1 + 10) &
          (q.p_size.between(1, 5)) &
          (q.l_shipmode.isin(('air', 'air reg'))) &
          (q.l_shipinstruct == 'deliver in person'))

    q2 = ((q.p_brand == BRAND2) &
          (q.p_container.isin(('med bag', 'med box', 'med pkg', 'med pack'))) &
          (q.l_quantity >= QUANTITY2) &
          (q.l_quantity <= QUANTITY2 + 10) &
          (q.p_size.between(1, 10)) &
          (q.l_shipmode.isin(('air', 'air reg'))) &
          (q.l_shipinstruct == 'deliver in person'))

    q3 = ((q.p_brand == BRAND3) &
          (q.p_container.isin(('lg case', 'lg box', 'lg pack', 'lg pkg'))) &
          (q.l_quantity >= QUANTITY3) &
          (q.l_quantity <= QUANTITY3 + 10) &
          (q.p_size.between(1, 15)) &
          (q.l_shipmode.isin(('air', 'air reg'))) &
          (q.l_shipinstruct == 'deliver in person'))

    q = q.filter([q1 | q2 | q3])
    q = q.aggregate(revenue=(q.l_extendedprice*(1-q.l_discount)).sum())
    return q
